src/librustc_resolve/late.rs

```diff
@@ -468,6 +468,19 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
                 }
             }));
 
+        // rust-lang/rust#61631: The type `Self` is essentially
+        // another type parameter. For ADTs, we consider it
+        // well-defined only after all of the ADT type parameters have
+        // been provided. Therefore, we do not allow use of `Self`
```

Comment:
I disagree with the phrasing (although the conclusion is similar) - see https://github.com/rust-lang/rust/issues/61631#issuecomment-538346412

In short, `Self` expands to/aliases `Adt<A, B, C, ...>`, referring to *all* the ADTs' parameters, and we shouldn't allow writing `Self` where we wouldn't allow `Adt<A, B, C, ...>`.